### PR TITLE
Deployment with Docker

### PR DESCRIPTION
**Pull request by ruioliveira02**

### CODE REVIEW

File: Dockerfile.dev [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+# Install build dependencies
+RUN apk add --no-cache build-base git python3
+
+# Install frontend dependencies
+RUN apk add --no-cache nodejs npm
+RUN npm install -g npx
+
+# Install imagemagick dependencies
+RUN apk add --no-cache file imagemagick
```
Comment (nelsonmestevao) [inline Dockerfile.dev:11]: This project does not uses imagemagick, nor nodejs.

File: Dockerfile.dev [hunk 2]
```diff
@@ -0,0 +1,22 @@
+FROM elixir:1.13-alpine
```
Comment (nelsonmestevao) [inline Dockerfile.dev:1]: I think we should start using 1.14 :point_up:

File: Dockerfile.prod [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+# Install build dependencies
+RUN apk add --no-cache build-base git python3
+
+# Install frontend dependencies
+RUN apk add --no-cache nodejs npm
+RUN npm install -g npx
+
+# Install imagemagick dependencies
+RUN apk add --no-cache file imagemagick
```
Comment (nelsonmestevao) [inline Dockerfile.prod:11]:
```suggestion
```

File: Dockerfile.prod [hunk 4]
```diff
@@ -0,0 +1,20 @@
+FROM elixir:1.13-alpine
```
Comment (nelsonmestevao) [inline Dockerfile.prod:1]:
```suggestion
FROM elixir:1.14-alpine
```

File: Dockerfile.prod [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+RUN apk add --no-cache file imagemagick
+
+# Install hex + rebar
+RUN mix local.hex --force && \
+    mix local.rebar --force
+
+
+WORKDIR /app
+
+CMD [ "sh", "-c", "mix deps.get; mix compile; mix assets.deploy; mix phx.server" ]
```
Comment (nelsonmestevao) [inline Dockerfile.prod:20]: This Dockerfile should be splitted in two sections, build the release, and then copy the files to a new container.

You should just use the phoenix generated Dockerfile, `mix phx.gen.release --docker`.

Learn more [here](https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Gen.Release.html).

File: config/runtime.exs [hunk 6]
```diff
@@ -55,12 +55,7 @@ if config_env() == :prod do
     client_id: spotify_client_id,
     client_secret: spotify_client_secret
 
-  database_path =
-    System.get_env("DATABASE_PATH") ||
-      raise """
-      environment variable DATABASE_PATH is missing.
-      For example: /etc/karaokium/karaokium.db
-      """
+  database_path = "/app/priv/repo/databases/karaokium_prod.db"
```
Comment (nelsonmestevao) [inline config/runtime.exs:58]: If this is a fixed variable, it should be moved to the compile time config `prod.exs`. Only keep dynamic variables in `runtime.exs`.

File: linux.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+version: "3"
+
+services:
```
Comment (nelsonmestevao) [inline linux.yml:3]: You are missing a `darwin.yml` for our MacOS friends.

Comment (nelsonmestevao) [reply linux.yml:3]: And this config is not compatible with the `ports` field.

File: Dockerfile.dev (hunk 2 above)
Comment (nelsonmestevao) [inline Dockerfile.dev:1]: I don't see any value in adding a Dockerfile for development. There's nothing to set up, and it makes much harder to work with mix.

File: .tool-versions [hunk 8]
```diff
@@ -1,2 +1,2 @@
-erlang 25.0.1
-elixir 1.13.4-otp-25
+erlang 25.1.2
```
Comment (nelsonmestevao) [inline .tool-versions:1]:
```suggestion
erlang 25.3
```

File: .tool-versions [hunk 9]
```diff
@@ -1,2 +1,2 @@
...
+elixir 1.14.0-otp-25
```
Comment (nelsonmestevao) [inline .tool-versions:2]:
```suggestion
elixir 1.14.3-otp-25
```

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+# This file is based on these images:
+#
+#   - https://hub.docker.com/r/hexpm/elixir/tags - for the build image
+#   - https://hub.docker.com/_/debian?tab=tags&page=1&name=bullseye-20210902-slim - for the release image
+#   - https://pkgs.org/ - resource for finding needed packages
+#   - Ex: hexpm/elixir:1.13.4-erlang-25.1.2-debian-bullseye-20210902-slim
+#
+ARG ELIXIR_VERSION=1.13.4
+ARG OTP_VERSION=25.1.2
+ARG DEBIAN_VERSION=bullseye-20210902-slim
```
Comment (nelsonmestevao) [inline Dockerfile:17]: Try to use the latest available image 💪

File: config/prod.exs [hunk 11]
```diff
@@ -49,3 +49,9 @@ config :logger, level: :info
 #       force_ssl: [hsts: true]
 #
 # Check `Plug.SSL` for all available options in `force_ssl`.
+
+database_path = "/app/priv/repo/databases/karaokium_prod.db"
+
+config :karaokium, Karaokium.Repo,
+  database: database_path,
+  pool_size: String.to_integer(System.get_env("POOL_SIZE") || "5")
```
Comment (nelsonmestevao) [inline config/prod.exs:57]: These dynamic field should be at runtime.exs

File: config/stg.exs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+#
+# We also recommend setting `force_ssl` in your endpoint, ensuring
+# no data is ever sent via http, always redirecting to https:
+#
+#     config :karaokium, KaraokiumWeb.Endpoint,
+#       force_ssl: [hsts: true]
+#
+# Check `Plug.SSL` for all available options in `force_ssl`.
+
+database_path = "/app/priv/repo/databases/karaokium_prod.db"
```
Comment (nelsonmestevao) [inline config/stg.exs:53]: You are using the same database for staging and production.

I think you should do string interpolation with config_env() instead.

File: docker-compose.prod.yml [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    env_file: .env.prod
+    environment:
+      MIX_ENV: ${MIX_ENV:-prod}
+    build:
+      context: .
+      dockerfile: Dockerfile
+    volumes:
+      - ./:/app
+      - /app/_build
+      - /app/deps
```
Comment (nelsonmestevao) [inline docker-compose.prod.yml:15]:
```suggestion
```

File: docker-compose.stg.yml [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    env_file: .env.stg
+    environment:
+      MIX_ENV: ${MIX_ENV:-stg}
+    build:
+      context: .
+      dockerfile: Dockerfile
+    volumes:
+      - ./:/app
+      - /app/_build
+      - /app/deps
```
Comment (nelsonmestevao) [inline docker-compose.stg.yml:15]:
```suggestion
```

File: rel/overlays/bin/migrate [hunk 15]
```diff
@@ -1,3 +1,3 @@
 #!/bin/sh
-cd -P -- "$(dirname -- "$0")/.."
```
Comment (nelsonmestevao) [inline rel/overlays/bin/migrate:2]: I think you are breaking relative paths here.

Why the change?

**Review by nelsonmestevao — CHANGES_REQUESTED** [review]

File: rel/overlays/bin/migrate (hunk 15 above)
Comment (ruioliveira02) [reply rel/overlays/bin/migrate:2]: It was auto generated by `mix phx.gen.release --docker`
